File: validation/poc-00/java-candidate/src/main/java/dev/vra/poc00/domain/DomainFailure.java
package dev.vra.poc00.domain;

/** Expected domain rejections; adapters choose transport status codes. */
public abstract sealed class DomainFailure extends RuntimeException {
    private DomainFailure(String message) { super(message); }

    public static final class InsufficientStock extends DomainFailure {
        public InsufficientStock() { super("Insufficient stock."); }
    }
    public static final class InvalidQuantity extends DomainFailure {
        public InvalidQuantity() { super("Quantity must be positive."); }
    }
    public static final class InvalidTransition extends DomainFailure {
        public InvalidTransition() { super("Order transition is not allowed."); }
    }
    public static final class InvalidIdempotencyKey extends DomainFailure {
        public InvalidIdempotencyKey() { super("Idempotency key must contain 1 to 128 characters."); }
    }
    public static final class CurrencyMismatch extends DomainFailure {
        public CurrencyMismatch() { super("Currencies must match."); }
    }
    public static final class UnknownSku extends DomainFailure {
        public UnknownSku() { super("SKU was not found."); }
    }
}

File: validation/poc-00/kotlin-candidate/src/main/kotlin/dev/vra/poc00/kotlin/domain/DomainFailure.kt
package dev.vra.poc00.kotlin.domain

/** Expected domain rejections; HTTP status and messages remain adapter concerns. */
sealed class DomainFailure(message: String) : RuntimeException(message) {
    class InsufficientStock : DomainFailure("Insufficient stock.")
    class InvalidQuantity : DomainFailure("Quantity must be positive.")
    class InvalidTransition : DomainFailure("Order transition is not allowed.")
    class InvalidIdempotencyKey : DomainFailure("Idempotency key must contain 1 to 128 characters.")
    class CurrencyMismatch : DomainFailure("Currencies must match.")
    class UnknownSku : DomainFailure("SKU was not found.")
}
